File: CrowdAssist/app/src/main/java/th/ac/kmitl/it/crowdalert/component/RoundLinearLayout.java
package th.ac.kmitl.it.crowdalert.component;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Path;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.widget.LinearLayout;

import th.ac.kmitl.it.crowdalert.R;

public class RoundLinearLayout extends LinearLayout {
    private int radius = 20;
    private Path path;
    public RoundLinearLayout(Context context) {
        super(context);
        init();
    }

    public RoundLinearLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.RoundLinearLayout);
        radius = a.getInt(R.styleable.RoundLinearLayout_radius, 20);
        a.recycle();
        init();
    }

    public RoundLinearLayout(Context context, AttributeSet attrs,
                             int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }
    private void init(){
        path = new Path();
    }
    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        path.reset();
        RectF rect = new RectF();
        rect.set(0, 0, w, h);
        path.addRoundRect(rect, radius, radius, Path.Direction.CW);
        path.close();
    }
    @Override
    public void draw(Canvas canvas) {
        int save = canvas.save();
        canvas.clipPath(path);
        super.draw(canvas);
        canvas.restoreToCount(save);
    }
}

File: CrowdAssist/app/src/main/java/th/ac/kmitl/it/crowdassist/component/RoundLinearLayout.kt
package th.ac.kmitl.it.crowdassist.component

import android.content.Context
import android.graphics.Canvas
import android.graphics.Path
import android.graphics.RectF
import android.util.AttributeSet
import android.widget.LinearLayout
import th.ac.kmitl.it.crowdassist.R
import kotlin.properties.Delegates

class RoundLinearLayout:LinearLayout{
    var radius = 20
    var path : Path by Delegates.notNull<Path>()
    constructor(context : Context): super(context)
    constructor(context: Context, attrs : AttributeSet):super(context, attrs){
        val a = context.obtainStyledAttributes(attrs, R.styleable.RoundLinearLayout)
        radius = a.getInt(R.styleable.RoundLinearLayout_radius, 20)
        a.recycle()
    }
    constructor(context : Context, attrs : AttributeSet, defStyleAttr : Int):super(context, attrs, defStyleAttr)
    init {
        path = Path()
    }
    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        path.reset()
        val rect = RectF()
        rect.set(0f, 0f, w.toFloat(), h.toFloat())
        path.addRoundRect(rect, radius.toFloat(), radius.toFloat(), Path.Direction.CW)
        path.close()
    }

    override fun draw(canvas: Canvas) {
        val save = canvas.save()
        canvas.clipPath(path)
        super.draw(canvas)
        canvas.restoreToCount(save)
    }
}